### PR TITLE
Patternlab/dp 20949 google map swap poc (merge order: 3)

### PR DESCRIPTION
**Pull request by clairesunstudio**
Swap out google maps with leaflet maps

Before:
Screen Shot 2021-04-09 at 4 49 09 PM

After:
Screen Shot 2021-04-09 at 4 49 17 PM

Before:
Screen Shot 2021-04-09 at 4 47 09 PM

After:
Screen Shot 2021-04-09 at 4 47 17 PM
---
### Data Changes:
```
leafletMap: {
  isStatic  // static --> isStatic
  hideAttribution // new field
  map: {}, // no changes
  markers: [] // no changes
 }
```
----

##### molecules:
~~http://localhost:3001/?p=molecules-leaflet-map~~ --> http://localhost:3001/?p=molecules-map-leaflet (interactive) 
~~http://localhost:3001/?p=molecules-leaflet-map-static~~ -->  http://localhost:3001/?p=molecules-map-leaflet-static (static)
##### organisms:
http://localhost:3001/?p=organisms-location-banner (static)
http://localhost:3001/?p=organisms-mapped-locations (interactive)
##### templates and pages:
http://localhost:3001/?p=templates-location (static)
http://localhost:3001/?p=pages-location-park-content (interactive)
http://localhost:3001/?p=pages-organization (interactive)
http://localhost:3001/?p=pages-service-flexible-link-groups (interactive)
http://localhost:3001/?p=templates-location (static)
http://localhost:3001/?p=pages-location-general-content (static)
http://localhost:3001/?p=pages-location-park-content (interactive)
http://localhost:3001/?p=pages-organization (interactive)
http://localhost:3001/?p=pages-organization-boards (interactive)
http://localhost:3001/?p=templates-services (interactive)
http://localhost:3001/?p=pages-service-flexible-link-groups (interactive)
http://localhost:3001/?p=pages-detail-for-service-howto-location (interactive)

---

To-do and questions:

1. [x] markers on mobile sizes are cut off slightly (might be related to media query css)
Screen Shot 2021-04-09 at 4 46 03 PM
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-20949-google-map-swap-poc/?p=molecules-leaflet-map-static 

2. [x] is there a way in twig to not remap the map data? (it causes the whole data object to be undefined in JS if certain fields are missing)

```
  <script>
    // Make leafletMap data object available in leaflet map module js.
    var ma = ma || {};
    ma.leafletMapData = ma.leafletMapData || [];
    ma.leafletMapData.push({{leafletMap}});
  </script>
```
Question: in `leaflet-map.twig`, why can't we do something like this ^
Note: The way that we currently mapping the data object is making all fields required and also not allowing boolean values
google-map.twig --> https://github.com/massgov/mayflower/blob/develop/packages/patternlab/styleguide/source/_patterns/02-molecules/google-map.twig#L21-L55
leaflet-map.twig -->https://github.com/massgov/mayflower/blob/develop/packages/patternlab/styleguide/source/_patterns/02-molecules/google-map.twig#L21-L55

3. [x] might need to adjust marker position (compare to google map)
4. [x] set min zoom (max bounds)?  Don't allow zoom or pane outside of the max bounds
Screen Shot 2021-04-13 at 6 50 47 PM
Screen Shot 2021-04-13 at 6 53 34 PM

4. [x] clean up code 
5. [x] additional improvements on static maps
  > Previously when we use google maps in location banner, we chose to render a static image of the map to reduce the requests to Google maps API due to cost reasons. Now we don't have the cost concerns, we have the opportunities to make the decision fully based on user experience.
The map is static – not allowing users to zoom out and pane around limits the usability of the map, and making it mostly only decorative. 
Screen Shot 2021-04-15 at 10 19 54 AM

We want to allow interactions on the maps, and move the directions link outside of the map so it doesn't collide with the interactions on the map.

### CODE REVIEW

**Comment by clairesunstudio** [comment]
I noticed leaflet library has a focus style issue that is not meeting accessibility requirements
related issue --> https://github.com/Leaflet/Leaflet/issues/6986
related fix --> https://github.com/Leaflet/Leaflet/pull/7259

waiting on upstream release ^

File: packages/patternlab/styleguide/source/assets/js/modules/mapLeaflet.js [hunk 1]
```diff
@@ -69,7 +69,7 @@ export default (function (window,document) {
 
       // custom marker icon 
       var markerIcon = L.icon({
-        iconUrl: `${ma.iconPath}marker-blue.svg`,
+        iconUrl: `${ma.iconPath}/marker-blue.svg`,
```
Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/mapLeaflet.js]: @vordude this changed broke the path in patternlab
https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-20949-google-map-swap-poc/?p=pages-location-general-content
Screen Shot 2021-04-28 at 2 16 10 PM

Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/mapLeaflet.js]:
```suggestion
        iconUrl: `${ma.iconPath}marker-blue.svg`,

```

Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/mapLeaflet.js]:
```suggestion
        iconUrl: `${ma.iconPath}marker-blue.svg`,

```
fix marker in patternlab

Comment (clairesunstudio) [inline packages/patternlab/styleguide/source/assets/js/modules/mapLeaflet.js]:
```suggestion
        iconUrl: `${ma.iconPath}marker-blue.svg`,

```

**Review by clairesunstudio — COMMENTED** [review]
fix marker in patternlab

**Comment by vordude** [comment]
This looks good at this point. Feature 2 has the code from this and openmass.

**Comment by clairesunstudio** [comment]
Feature2 looks good! I see a bug in the location listing in patternlab https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-20949-google-map-swap-poc/?p=organisms-location-listing
I will look into it
Screen Shot 2021-04-29 at 5 19 06 PM

**Comment by clairesunstudio** [comment]
Fixes tested in safari and chrome:

**before:**
safari-after
**after:**
safari-before

---
**before:**
bounds-before
**after:**
bounds-after

**Comment by clairesunstudio** [comment]
- Fixing popup being cropped on certain mobile devices (iPhone 7, iphone8, Galaxy S20, Galaxy S21)
- Repositioned zoom controls to unblock popup info
Screen Shot 2021-05-04 at 3 48 23 PM

Screen Shot 2021-05-04 at 5 14 04 PM

**Comment by clairesunstudio** [comment]
Prevent scolling/swiping ambiguity
Only enable scroll zoom and pane if map is in focus, and disable after user click outside of the map 
Use icons to track interactivity state

map-scroll-fix-state

**Comment by clairesunstudio** [comment]
Code cleanup needed after external review

**Review by ygannett — APPROVED** [review]
Didn't merge after the conversation with Joe about the scrolling feature/behavior.

**Comment by clairesunstudio** [comment]
- [x] Remove lock icon
- [x] Add message "XX Click or tap INSIDE map to move the map" and after you click in, switch to "XX Click or tap OUTSIDE map to scroll the page".   Try to find best location.
- [x] If someone clicks a marker, make sure we unlock the map so that when they close the marker, they can move the map. (works with click, tap, and tab on marker and popup close button)

Tested on Chrome, Sarafi, iPhone 7,8 and Galaxy S9+ (browserstack)
---

map-scroll-fix-mobile

**Comment by clairesunstudio** [comment]
Map bugs:
- [x] recenters on scroll on mobile (iPhone 8, 12)
- [x] map on location pages hard to pane on mobile
- [ ] On mobile, message are not being updated correctly when map is active… (https://mayflower.digital.mass.gov/patternlab/b/patternlab/DP-20949-google-map-swap-poc/patterns/05-pages-location-park-content/05-pages-location-park-content.html) This is happening when multiple map is on one page

**Comment by clairesunstudio** [comment]
Help wanted!

Because leaflet markers are not added as part of the maps in events. I had to fake onBlur by adding event listeners on document. The event listener is added each time initMaps is called, when there are multiple maps on a page, the onclick is overriden by the last initMaps hence the mapWrapper is not the correct one for the other maps.
```
      document.onclick = (e) => {
        const { target } = e;
        // console.log(target)
        customBlur(target)
      };

      document.onkeyup = (e) => {
        if (e.key == 'Tab') {
          const target = document.activeElement;
          customBlur(target)
        }
      }
```

**Comment by clairesunstudio** [comment]
The latest build is brought in https://github.com/massgov/openmass/pull/707 and deployed to feature2

Page examples for x-browser/x-device testing:
https://massgovfeature2.prod.acquia-sites.com/orgs/massachusetts-registry-of-motor-vehicles 
https://massgovfeature2.prod.acquia-sites.com/locations/auburn-aaa-limited-rmv-services

**Pull request merged**